[material-ui][docs] Revise the Snackbar page (#39298)
mui/material-ui: docs/data/material/components/snackbars/snackbars.md

# Snackbar

> <p class="description">Also known as a toast, snackbars are used for brief notifications of a process that has been performed or will be performed.</p>

Reviewer: ```suggestion
<p class="description">Snackbars (also known as toasts) are used for brief notifications of processes that have been or will be performed.</p>
```

{{"component": "modules/components/ComponentLinkHeader.js"}}

## Introduction

```jsx
import Snackbar from '@mui/material/Snackbar';

> ```

Reviewer: I think we should move the import down to the Basics section, as we've done with the Joy and Base docs. I could imagine an intro playground here in a future version of the Material UI docs!

> The Snackbar provides users updates on an app's processes, temporarily appearing on the screen.

Reviewer: I think "floats above the UI" and "non-critical updates" are important details here for differentiating a Snackbar from other similar components, based on something I saw in a Stack Overflow question recently

```suggestion
The Snackbar component appears temporarily and floats above the UI to provide users with (non-critical) updates on an app's processes.
```

> Inspired by Google Keep, this demo shows a basic snackbar with a text element and two actions.

Reviewer: ```suggestion
The demo below, inspired by Google Keep, shows a basic Snackbar with a text element and two actions:
```

{{"demo": "SimpleSnackbar.js"}}

> 

Reviewer: How about a note on usage here? I think it's helpful to compare/contrast with Alerts and Dialogs so users can understand the appropriate use cases.

```suggestion
:::info
Snackbars differ from [Alerts](/material-ui/react-alert/) in that Snackbars have a fixed position and a high z-index, so they're intended to break out of the document flow; Alerts, on the other hand, are usually part of the flow—except when they're [used as children of a Snackbar](#use-with-alert).

Snackbars also from differ from [Dialogs](material-ui/react-dialog) in that Snackbars are not intended to convey _critical_ information or block the user from interacting with the rest of the app; Dialogs, by contrast, require input from the user before they can be dismissed.
:::

```

## Basics

### Position

> Use the `anchorOrigin` prop to control the snackbar position on the screen.

Reviewer: ```suggestion
Use the `anchorOrigin` prop to control the Snackbar's position on the screen.
```

{{"demo": "PositionedSnackbar.js"}}

### Content

```jsx
import SnackbarContent from '@mui/material/SnackbarContent';
```

> Use the Snackbar Content component for adding text and actions to the snackbar.

Reviewer: ```suggestion
Use the Snackbar Content component to add text and actions to the Snackbar.
```

{{"demo": "LongTextSnackbar.js"}}

### Transitions

You can use the `TransitionComponent` prop to change the transition of the Snackbar from [Grow](/material-ui/transitions/#grow) (the default) to others such as [Slide](/material-ui/transitions/#slide).

{{"demo": "TransitionsSnackbar.js"}}

## Customization

### Use with Alerts

> Use an alert inside a snackbar for messages that communicate a certain severity.

Reviewer: ```suggestion
Use an Alert inside a Snackbar for messages that communicate a certain severity.
```

{{"demo": "CustomizedSnackbars.js"}}

### Use with Floating Action Buttons

> If you're using a [Floating Action Button](/material-ui/react-floating-action-button/) on mobile, Material Design recommends positioning any snackbar right above it.
> The demo below shows you how to do that.

Reviewer: ```suggestion
If you're using a [Floating Action Button](/material-ui/react-floating-action-button/) on mobile, Material Design recommends positioning Snackbars directly above it, as shown in the demo below:
```

{{"demo": "FabIntegrationSnackbar.js", "iframe": true, "maxWidth": 400}}

## Common examples

### Consecutive Snackbars

> This demo shows how to show multiple snackbars without stacking them by using a consecutive animation.

Reviewer: ```suggestion
This demo shows how to display multiple Snackbars without stacking them by using a consecutive animation.
```

{{"demo": "ConsecutiveSnackbars.js"}}

> ## Third-party intergrations

Reviewer: ```suggestion
## Third-party integrations
```

### notistack

![stars](https://img.shields.io/github/stars/iamhosseindhv/notistack.svg?style=social&label=Star)
![npm downloads](https://img.shields.io/npm/dm/notistack.svg)

> With an imperative API, [notistack](https://github.com/iamhosseindhv/notistack) allows you to vertically stack multiple Snackbars without having to handle their open and close states.

Reviewer: ```suggestion
With an imperative API, [notistack](https://github.com/iamhosseindhv/notistack) lets you vertically stack multiple Snackbars without having to handle their open and close states.
```

Even though this is discouraged in the Material Design guidelines, it is still a common pattern.

{{"demo": "IntegrationNotistack.js", "defaultCodeOpen": false}}

## Accessibility

Here are a few tips to make sure you have an accessible Snackbar component:

> - Based on the [WAI-ARIA guidelines](https://www.w3.org/TR/wai-aria-1.1/#alert), the Snackbar doesn't auto-hide by default.

Reviewer: ```suggestion
- According to the [WAI-ARIA guidelines](https://www.w3.org/TR/wai-aria-1.1/#alert), Snackbars shouldn't auto-hide by default.
```

>   If you use the `autoHideDuration` prop, make sure to [provide sufficient time](https://www.w3.org/TR/UNDERSTANDING-WCAG20/time-limits.html) for the user to process the information displayed on it.

Reviewer: this page doesn't mention this prop elsewhere, does it? Maybe we should add a section and a demo to show it off? But then again, it seems like we should probably discourage devs from actually using it (according to WAI-ARIA), so, maybe not... ?

Author: Uhm... good question. Interestingly, though, and not sure if I misunderstood it, I didn't feel like the Aria link there mentions precisely what we're recommending _not_ to do. A specific part made me confused:

> "Since alerts are not required to receive focus, content authors SHOULD NOT require users to close an alert."

Does this mean that it's actually okay to auto-hide it, given it doesn't require focus, therefore not requiring user action? With that said, I'd be ok not to add anything now, but it made me question if this recommendation is even correct (thus, worth adding a demo/section).

Reviewer: I did some more reading on the Snackbar / Toast pattern, and it seems that because they're not intended to convey critical information - more just to confirm that an expected action has taken place (email sent, form submitted) - they should automatically be dismissed with no action taken by the user. MD2 suggests 4 to 10 seconds: https://m2.material.io/components/snackbars

So I do think we should add a section to demonstrate the `autoHideDuration` prop, and maybe move the "provide sufficient time" link to that section and cut the link about how it shouldn't auto-hide. What do you think?

> - All Snackbars are automatically dismissed by pressing <kbd class="key">Escape</kbd>. If you have multiple instances appearing at the same time and want <kbd class="key">Escape</kbd> to dismiss only the oldest currently open one, call `event.preventDefault` in the `onClose` prop.

Reviewer: ```suggestion
- The user should be able to dismiss Snackbars by pressing <kbd class="key">Escape</kbd>. If there are multiple instances appearing at the same time and you want <kbd class="key">Escape</kbd> to dismiss only the oldest one that's currently open, call `event.preventDefault` in the `onClose` prop.
```

# Snackbar

<p class="description">Also known as a toast, snackbars are used for brief notifications of a process that has been performed or will perform.</p>

{{"component": "modules/components/ComponentLinkHeader.js"}}

> ## Basics

Author: Similar to the comment below — as per the designed Base UI docs structure (https://github.com/mui/material-ui/pull/38529), I guess we'll be changing this very first demo to _Introduction_ instead of either Basics or Simple (just fixed it). cc @samuelsycamore for further guidance!

> ## Basics

Author: Resolving this one to concentrate the discussion on the comment below!

```jsx
import Snackbar from '@mui/material/Snackbar';
```

Inspired by Google Keep, the demo below shows a basic snackbar with a text element and two actions.

{{"demo": "SimpleSnackbar.js"}}

> ## Customization

Reviewer: "Component" as a header really only makes sense to me in the context of Base UI (where you might be looking for info on either components or hooks), which is why we dropped it from the Joy UI docs. I'd say the Joy docs provide a better template for what we should aim for in the Material UI docs.

@mbrookes does bring up a good point, which is: where do we draw the line between what is "basic" stuff vs "customization"? So far in the Base and Joy docs, we've mostly put anything that isn't the _default_ behavior or style under "Customization," but does that really make sense? For instance, in the Joy docs—is it really "customization" to apply global variants, or is that actually "basic" stuff?

Another thing the Joy UI docs do that I really like is the "Common examples" sections (which I see you've added here), and I do think Material UI needs the same thing.

So it sounds like we need three h2s on (most) component pages for the following sections:

- Basics
- Customization
- Common examples

I would propose breaking them up like this:

- Basics: default behavior and styling, plus how to apply built-in props
- Customization: adjusting styles and behaviors beyond merely applying props; use with other components
- Common examples: fleshed out demos showing the full range of options available to the components, often recreating well known bits of UI / popular design systems

Does that make sense?

So with this proposed format, the Snackbar page would look more or less like this:

```
# Snackbar

## Introduction

## Basics
### Position
### Content
### Transitions

## Customization
### Use with Alerts
### Use with Floating Action Buttons

## Common examples
### Consecutive Snackbars (or maybe this belongs in Customization?)

## Third-party integrations
### notistack

## Accessibility

## Anatomy

## API
```

### Position

Use the `anchorOrigin` prop to control the snackbar position on the screen.

{{"demo": "PositionedSnackbar.js"}}

### Content

```jsx
import SnackbarContent from '@mui/material/SnackbarContent';
```

Use the Snackbar Content component for adding text and actions to the snackbar.

{{"demo": "LongTextSnackbar.js"}}

> ### Usage with Alerts

Reviewer: I'm cool with either term as long as apply it consistently. As for capitalization, I'd keep "Alerts" capitalized since it refers to our component specifically (rather than the general concept of _an_ alert).

# Snackbar

<p class="description">Snackbars (also known as toasts) are used for brief notifications of processes that have been or will be performed.</p>

{{"component": "modules/components/ComponentLinkHeader.js"}}

## Introduction

The Snackbar component appears temporarily and floats above the UI to provide users with (non-critical) updates on an app's processes.
The demo below, inspired by Google Keep, shows a basic Snackbar with a text element and two actions:

{{"demo": "SimpleSnackbar.js"}}

> :::info

Author: Yeah, I like the idea for components that have a bigger need for clarification, the Snackbar/Alert/Dialog stack being a good example. Maybe Menu/Popover/Popper would be another one. 👌

> ### Usage

Author: Yeah, it definitely feels like a case-by-case sort of thing! It'd be good to document this possibility on that issue, though!

> Snackbars differ from [Alerts](/material-ui/react-alert/) in that Snackbars have a fixed position and a high z-index, so they're intended to break out of the document flow; Alerts, on the other hand, are usually part of the flow—except when they're [used as children of a Snackbar](#use-with-alert).

Author: Good catch!

> Snackbars also from differ from [Dialogs](/material-ui/react-dialog/) in that Snackbars are not intended to convey _critical_ information or block the user from interacting with the rest of the app; Dialogs, by contrast, require input from the user before they can be dismissed.

Author: Keep these ones coming!

## Basics

> 

Author: I agree! cc @samuelsycamore so we add it to the template 👍

{{"demo": "TransitionsSnackbar.js"}}

### Automatic dismiss

> Use the `autoHideDuration` prop to specify an amount of time (in milliseconds) for the close function to be called, and thus, for the Snackbar to be automatically dismissed.

Author: Should it be then just (removed the passive voice, too):

> Use the `autoHideDuration` prop to specify an amount of time (in milliseconds) for automatically dismissing the Snackbar.

A reason why added the part about the close function is that, trimming it that way, gives me the impression that just by adding this prop and setting a time, the Snackbar will automatically close. There’s no “automatic” close, as in that’s not built-in; folks still need to put in a handle close function there. But maybe that’s already implicit, though?

> Use the `autoHideDuration` prop to specify an amount of time (in milliseconds) for the close function to be called, and thus, for the Snackbar to be automatically dismissed.

Author: Sounds good to me!

  );
}
```

## Anatomy

> The Snackbar component is composed of a root <div> that ... to complement:

Reviewer: ```suggestion
The Snackbar component is composed of a root `<div>` that houses interior elements like the Snackbar Content and other optional components (such as buttons or decorators).
```

```html

> <div class="MuiSnackbar-root">
>   <!-- snackbar content goes here -->
> </div>

Reviewer: Here's the basic anatomy - I'm a little confused about what's going on here, honestly 😅 not sure why this Paper is involved or why the content is wrapped in two layers of divs but I guess it is what it is!

```suggestion
<div role="presentation" class="MuiSnackbar-root>
  <div class="MuiPaper-root MuiSnackbarContent-root" role="alert">
    <div class="MuiSnackbarContent-message">
      <!-- Snackbar content goes here -->
    </div> 
  </div>
</div>
```
